### ISSUE TITLE
RPC and operation data callback

### ISSUE THREAD

**Issue by iljatab** [sysrepo/sysrepo#1963]
$ sysrepoctl --version
sysrepoctl - sysrepo YANG schema manipulation tool, compiled with libsysrepo v1.4.55 (SO v5.4.14)

$ yanglint --version
yanglint SO 1.7.9

$ netopeer2-server -V
netopeer2-server 1.1.25
compile time: May  4 2020, 08:45:56

Hi Michal,

I have a strange issue, not sure where is the bug.

In our application we have some RPC callbacks, installed by sr_rpc_subscribe().
Also there are operational state callbacks, installed by sr_oper_get_items_subscribe().
At sysrepo level there just actually just one RPC callback and one operational state callback.
Inside each of the two callbacks we do our own dispatching to actual YANG module instrumentation.
Both callbacks use same subscription with SR_SUBSCR_CTX_REUSE.

The issue: it looks that for all RPC calls (done by `netopeer-cli` `user-rpc` command) our RPC callback is not called. Instead the operational state data callback in called with some completely unrelated arguments.
I cannot tell when this issue started to happen. We have just a handful of RPCs and we don't yet have automated tests for those. But it definitely worked after our migration to sysrepo 1.x. And we didn't change anything on our RPC stuff for months.

Do you have any ideas what could be wrong?
Or should we look for the bug on our side?

Ilja.

**Comment by michalvasko** [comment]
Hi Ilja,
based on the description I cannot help you at all. But *sysrepo* has fairly detailed messages about every callback it calls (`INF` level or `3`) so you want to check those.

Regards,
Michal

**Comment by iljatab** [comment]
Hi Michal,

I got some logs but still cannot understand what's going on.
Please find a tarball attached with:
1. YANG model where RPC is defined
2. XML file with RPC request, used with netopeer2-cli
3. netopeer2-server log with `-d 2`, it covers exactly the RPC call

As you can see, the RPC has nothing to do with openconfig-interfaces model.
However the whole log is filled with messages like:
```
[INF]: SR: Published event "operational" "/openconfig-interfaces:interfaces/interface/state" with ID 229.
[ERR]: SR: basic_string::_M_construct null not valid

```
The `SR: basic_string::_M_construct null not valid` messages are actually originated from our operational state callback handler in our application - it is called instead of our rpc callback handler with garbage input.

The netopeer2-cli gets:
```
> user-rpc --content oam_test_start.xml
ERROR
        type:     application
        tag:      operation-failed
        severity: error
        message:  basic_string::_M_construct null not valid

        type:     application
        tag:      operation-failed
        severity: error
        message:  basic_string::_M_construct null not valid

        type:     application
        tag:      operation-failed
        severity: error
        message:  basic_string::_M_construct null not valid
```

Could you take a look?

Ilja.

**Comment by iljatab** [comment]
Hmmm... looks like I forgot the attachment.
[logs-1.tar.gz](https://github.com/sysrepo/sysrepo/files/4584941/logs-1.tar.gz)

**Comment by michalvasko** [comment]
Hi Ilja,
well, in general it makes sense because before an RPC can be sent, it is validated against the `operational` datastore. And since your RPC includes a `leafref` to `openconfig-interfaces`, its data are also being read. So, the only question remains what do you mean exactly by saying that the operational callbacks get some garbage input? I also see in the output that every operational callback seems to get called several times, which does not seem correct to me but it depends.

Regards,
Michal

**Comment by iljatab** [comment]
Hi Michal,

Actually I cannot understand why operational data callback should be called at all.
Yes, there is a leafref to interface name in RPC - but the leafref points to configuration part of the openconfig-interfaces.yang. Not to operational data.

IMHO, all is needed to validate leafref in RPC is to check in the sysrepo database for openconfig-interfaces if such interface name exists. No need to pull out operational data. Also, as I can see, sysrepo likely tries to get operational data for **all** existing interfaces - therefore we see so many messages.

I remember it worked just fine earlier :-(

Ilja.

**Comment by michalvasko** [comment]
Hi Ilja,
based on the [NMDA RFC](https://tools.ietf.org/html/rfc8342#section-6.2) the `operational` datastore is always used for validation including finding targets of any leafrefs. But the text is quite short so I will write to the netmod mailing to make sure I understood it correctly.

Also, what do you mean it worked fine earlier? I believe it works this way for quite some time now. And you are right that it tries to get the whole operational datastore. That may perhaps be avoided so I will take a look at that later, it is just an optimization.

Regards,
Michal

**Comment by iljatab** [comment]
Thanks for clarification but I still am lost.

OK, leafref in RPC should be validated against `operational` datastore.
But how is it directly related to the operational state callbacks?
I might be wrong but it doesn't seem to me that we must to provide an operational data callback for configurable elements.
I would expect that configurable leafs are populated in operational datastore automatically from the configuration datastore, without the need to provide a callback for each and every leaf.

For `openconfig-interfaces` YANG model we call `sr_oper_get_items_subscribe()` to subscribe to following `path` elements in module `openconfig-interfaces`:
```
/openconfig-interfaces:interfaces/interface/state
/openconfig-interfaces:interfaces/interface/state/counters
/openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/state
/openconfig-interfaces:interfaces/interface/openconfig-if-ethernet:ethernet/openconfig-if-ethernet:state
/openconfig-interfaces:interfaces/interface/openconfig-if-ethernet:ethernet/openconfig-if-ethernet:state/counters
/openconfig-interfaces:interfaces/interface/openconfig-if-aggregate:aggregation/openconfig-if-aggregate:state
/openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/openconfig-vlan:vlan/openconfig-vlan:state
/openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/openconfig-if-ip:ipv4/openconfig-if-ip:state
/openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/openconfig-if-ip:ipv6/openconfig-if-ip:state
/openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/openconfig-if-ip:ipv4/openconfig-if-ip:addresses/openconfig-if-ip:address/openconfig-if-ip:state
/openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/openconfig-if-ip:ipv6/openconfig-if-ip:addresses/openconfig-if-ip:address/openconfig-if-ip:state
```
So basically we subscribe to the "no config" elements in the model to provide operational data.

However the `leafref` we are talking about (in RPC) points to `/openconfig-interfaces:interfaces/interface/name`.
In the YANG model this is a configurable leaf and it is out of scope of any operational data path we subscribe (as listed above).

So to us not calling these callbacks is not "just an optimization" but a major issue.
Getting state information for all interfaces would make RPC slower maybe 1000 times.

If needed I can also check why our data provider callback considers these calls as "garbage".

Ilja.

**Comment by michalvasko** [comment]
Hi Ilja,
you are right about everything you said and I think it does not disagree with what I said in any way. I may just clarify what I meant by "just an optimization", which is that even without it, it should function correctly. Yes, the consequence for your case is that a lot unnecessary data are retrieved, but, again, it does not affect correctness, only the performance. I have written to the mailing list so I hope someone will clarify how exactly is this supposed to work and only then I can decide how to improve/fix this.

Having said all that, the callbacks should be called correctly the way it functions now, so please try to check where is the problem.

Regards,
Michal

**Comment by iljatab** [comment]
Hi Michal,

Thanks for support!

I should admit that "garbage" in the callback was my fault :-)

The data provider callback is called with `request_xpath` equal to `NULL` and the callback was not prepared for that. When doing a regular `get`,  the 'request_xpath` never is `NULL` for any of our YANG models.

Now I can either:

1. If `request_xpath == NULL` just return from the callback immediately doing nothing.
I already tested this and it looks working - RPC proceeds as expected.
But not sure if this is a stable feature - can later `request_xpath` become non-NULL and break this logic?

or

2. Kindly ask you to implement optimization, that you mentioned.
Please note that because `request_xpath == NULL` we even don't have a chance to limit the tree to one interface (mentioned in RPC). The whole tree should be built and this is pretty expensive.

What do you think?

Ilja.

**Comment by michalvasko** [comment]
Hi Ilja,
I just checked and it is explicitly mentioned in the documentation that `request_xpath` can be `NULL` so nothing for me to improve there :) It is `NULL` for cases such as these when there is no specific user request but simply all the data should be returned.

1. `request_xpath` can definitely change and I believe there will still be cases when it will be `NULL` but the whole data tree actually will need to be returned so what you are proposing will not work.
2. Okay, I can take a look, it should not be too difficult.

Also, I have asked on the mailing list and got a confirmation that I understand it correctly, it really is validated against the `operational` datastore.

Regards,
Michal

**Comment by iljatab** [comment]
Thanks you for support Michal!

Yes, I understand the point better now.

We already implemented correct handling for `request_xpath == NULL` in our operational data callback and the issue went away. RPC started working. But, of course, it's slow.

Another option on our side is to avoid `leafter` entirely in RPC YANG model.
We can define interface name as a regular `string` in RPC and validate interface name in RPC callback.
This will be fast.

If you don't need to keep this ticket open for tracking purposes, please feel free to close it.

Thanks again!

Ilja.

**Comment by SheshavShalini** [comment]
Hi 
I am facing one issue, I am trying to read Operational data of device, but not able to get data. I read somewhere that if we subscribe, we should get operational data , so my question is, what command I should try to read operational data of device. I am using Netopeer 2. Please help

**Comment by michalvasko** [comment]
Please do not respond with unrelated queries to existing issues but rather create a new one. You did not provide much information but *operational* data should be part of `get` RPC reply. If you have any more questions, create a new issue. I will remove any new unrelated replies here.
